gui/default/syncthing/settings/advancedSettingsModalView.html

```diff
@@ -15,7 +15,7 @@ <h4 class="panel-title" translate tabindex="0">GUI</h4>
         <div id="guiConfig" class="panel-collapse collapse" role="tabpanel" aria-labelledby="guiHeading">
           <div class="panel-body less-padding">
             <form class="form-horizontal" role="form">
-              <div ng-repeat="(key, value) in advancedConfig.gui" ng-init="type = inputTypeFor(key, value)" ng-if="type != 'skip'" class="form-group">
+              <div ng-repeat="(key, value) in advancedConfig.gui" ng-if="inputTypeFor(key, value) != 'skip'" class="form-group">
```

Comment:
Hm. Yes and no? This is from commit 3d04e78a343f93b3ef1d20238643f4f515192bc8, and in its preceding commit, at least the `gui._webauthnCredentials` setting is incorrectly shown even though its `_`-prefixed name should hide it:

![image](https://github.com/user-attachments/assets/7afcfc5e-0460-4b76-8185-692fb8513763)

3d04e78a343f93b3ef1d20238643f4f515192bc8 fixes that... in that particular instance. Frustrating as it is, I haven't been able to reproduce the issue with any other setting, so I don't know why it's not working in that specific case. And as of commit 64edef4cbcbd54cfef03b72d5ff5166c342d1ee4 (where `gui._webauthnCredentials` no longer exists) I can't tell any difference with or without this change from commit 3d04e78a343f93b3ef1d20238643f4f515192bc8.

This is trivial to revert, so let me know if you'd like me to. If not, I'll leave it in.

Reply:
Sorry, that reordering comment was probably mistaken because I was looking at a partial change set.  So you're saying the `ng-init` stuff doesn't work as intended?  Isn't the `type` variable used locally for other things?

Reply:
> So you're saying the `ng-init` stuff doesn't work as intended?

It didn't work as intended for `gui._webauthnCredentials` (which no longer exists), but I haven't seen it fail in any other case.

> Isn't the `type` variable used locally for other things?

It is not (this is on `main`):
https://github.com/syncthing/syncthing/blob/5342bec1b75025355b0549dca4860113a3518b37/gui/default/syncthing/settings/advancedSettingsModalView.html#L18-L24

The would-be local uses all call `inputTypeFor()` again in-line.

Reply:
Well if the code doesn't use the local, then maybe because it doesn't work at all, as you witnessed.  Still, this is already a large PR this change would get buried in.  If you like, make another PR for it, or simply ignore it and revert in this PR.

Reply:
Ah - after merging https://github.com/emlun/syncthing/pull/7 this is now relevant again. On commit efd935a12bb6450ec1afd3671cdf8f4a79b5dd32 (with this change) the advanced settings are correct:

![image](https://github.com/user-attachments/assets/d01aae7d-1c86-489b-a5c0-0da90cd833a4)

But after reverting this change (`git checkout -p upstream/main -- gui/default/syncthing/settings/advancedSettingsModalView.html`), the `Object`-type `WebauthnState` attribute is incorrectly included:

![image](https://github.com/user-attachments/assets/c1554027-29a8-44a3-bdac-7fdb61bb61a8)

There's also a similar case in the "Folders" settings:

![image](https://github.com/user-attachments/assets/a9b41a40-698b-4010-b762-d43f86cf43e8)

but the cause of this one seems to that the proto definition type is `repeated FolderDeviceConfiguration`, so it's a list, so `inputTypeFor` returns `'list'` assuming that it's a list of strings, but it's a list of objects, so it should rather be skipped.

Anyway - since the motivation for the `ng-init` change doesn't seem to be observable without the added `WebauthnState` property, is it okay if I keep it bundled with this PR? Or would you still prefer I move it out to a separate PR, even if the problem isn't directly visible there?

Reply:
I think the basic problem is that `inputTypeFor()` does not filter out lists of objects.  So I tried to fix that, see https://github.com/syncthing/syncthing/pull/9743

Reply:
Nice! Unfortunately, that doesn't fix the problem with `gui.webauthnState`. It seems like `ng-init` and `ng-if` may not have a definite evaluation order, because the `ng-if` seems to work reliably for `folders.devices` and `gui.webauthnOrigins` (which gets hidden due to the [logic bug](https://github.com/syncthing/syncthing/pull/9743#discussion_r1780006512) in #9743 - this is incorrect but an interesting example) but not at all for `gui.webauthnState` for some reason. After merging #9743 into this branch, I need this additional diff in order to get `gui.webauthnState` to be correctly hidden:

```diff
diff --git a/gui/default/syncthing/settings/advancedSettingsModalView.html b/gui/default/syncthing/settings/advancedSettingsModalView.html
index ae92c0150..a39df7ef5 100644
--- a/gui/default/syncthing/settings/advancedSettingsModalView.html
+++ b/gui/default/syncthing/settings/advancedSettingsModalView.html
@@ -15,7 +15,7 @@
         <div id="guiConfig" class="panel-collapse collapse" role="tabpanel" aria-labelledby="guiHeading">
           <div class="panel-body less-padding">
             <form class="form-horizontal" role="form">
-              <div ng-repeat="(key, value) in advancedConfig.gui" ng-init="type = inputTypeFor(key, value)" ng-if="type != 'skip'" class="form-group">
+              <div ng-repeat="(key, value) in advancedConfig.gui" ng-init="type = inputTypeFor(key, value)" ng-if="inputTypeFor(key, value) != 'skip'" class="form-group">
                 <label for="guiInput{{$index}}" class="col-sm-4 control-label">{{key | uncamel}}&nbsp;<a href="{{docsURL('users/config#config-option-gui.')}}{{key | lowercase}}" target="_blank"><span class="fas fa-question-circle"></span></a></label>
                 <div class="col-sm-8">
                   <input ng-if="type == 'list'" id="guiInput{{$index}}" class="form-control" type="text" ng-model="advancedConfig.gui[key]" ng-list />
```

The [`ng-init` docs](https://code.angularjs.org/1.3.20/docs/api/ng/directive/ngInit) do say:

>The only appropriate use of `ngInit` is for aliasing special properties of [`ngRepeat`](https://code.angularjs.org/1.3.20/docs/api/ng/directive/ngRepeat), as seen in the demo below. Besides this case, you should use [controllers](https://code.angularjs.org/1.3.20/docs/guide/controller) rather than `ngInit` to initialize values on a scope.

so we probably shouldn't be using `ng-init` for this, and should instead use a custom controller if we really want to cache the `inputTypeFor` invocations.

Reply:
I just now re-confirmed that this change is still needed as of commit 54e2eb9b583d83900f6e9a55f5a7892b8dac4088 - reverting it makes `WebAuthn State: [object Object]` show up in the advanced settings again. This does however clash with the change in 896b857fc45b07bdbc1a81b1ba362863b5d2747f which now requires the `type` variable in the `ng-repeat` body. The variable seems to work fine within the body, but it seems like `ng-if` and `ng-init` have indeterminate evaluation order, so I reverted to using `inputTypeFor` in the `ng-if` check. This seems to reliably skip items as appropriate while keeping the correct input type.